# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/events/IMEStateManager.cpp
# Commit: c8cf9b256323
# Full Hash: c8cf9b256323da1c2f488ccef3a88369ffa3c051
# Author: Henri Sivonen <hsivonen@hsivonen.fi>
# Date: 2019-04-01 21:54:51
# Regressor Bug: 1524975
# File Overlap Count: 1
# Description:
#   Bug 1524975 - Split cross-process IME committing and blurring out of IMEStateManager::OnChangeFocusInternal(). r=masayuki
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D23655
# ==============================================================================

diff -r 508dd8a22136 -r c8cf9b256323 dom/events/IMEStateManager.cpp
--- a/dom/events/IMEStateManager.cpp	Mon Apr 01 16:45:00 2019 +0000
+++ b/dom/events/IMEStateManager.cpp	Mon Apr 01 07:22:54 2019 +0000
@@ -18,6 +18,7 @@
 #include "mozilla/TextComposition.h"
 #include "mozilla/TextEvents.h"
 #include "mozilla/Unused.h"
+#include "mozilla/dom/BrowserBridgeChild.h"
 #include "mozilla/dom/HTMLFormElement.h"
 #include "mozilla/dom/MouseEventBinding.h"
 #include "mozilla/dom/TabParent.h"
@@ -122,24 +123,12 @@
   }
 }
 
-static bool IsSameProcess(const TabParent* aTabParent1,
-                          const TabParent* aTabParent2) {
-  if (aTabParent1 == aTabParent2) {
-    return true;
-  }
-  if (!aTabParent1 != !aTabParent2) {
-    return false;
-  }
-  return aTabParent1->Manager() == aTabParent2->Manager();
-}
-
 StaticRefPtr<nsIContent> IMEStateManager::sContent;
 StaticRefPtr<nsPresContext> IMEStateManager::sPresContext;
 nsIWidget* IMEStateManager::sWidget = nullptr;
 nsIWidget* IMEStateManager::sFocusedIMEWidget = nullptr;
 StaticRefPtr<TabParent> IMEStateManager::sFocusedIMETabParent;
 nsIWidget* IMEStateManager::sActiveInputContextWidget = nullptr;
-StaticRefPtr<TabParent> IMEStateManager::sActiveTabParent;
 StaticRefPtr<IMEContentObserver> IMEStateManager::sActiveIMEContentObserver;
 TextCompositionArray* IMEStateManager::sTextCompositions = nullptr;
 InputContext::Origin IMEStateManager::sOrigin = InputContext::ORIGIN_MAIN;
@@ -180,30 +169,67 @@
 }
 
 // static
-void IMEStateManager::OnTabParentDestroying(TabParent* aTabParent) {
-  if (sFocusedIMETabParent == aTabParent) {
-    NotifyIMEOfBlurForChildProcess();
-  }
-
-  if (sActiveTabParent != aTabParent) {
-    return;
-  }
-
-  MOZ_LOG(sISMLog, LogLevel::Info,
-          ("OnTabParentDestroying(aTabParent=0x%p), "
-           "The active TabParent is being destroyed",
-           aTabParent));
-
-  // The active remote process might have crashed.
-  sActiveTabParent = nullptr;
-
-  // XXX: Need to disable IME?
-}
-
-// static
 void IMEStateManager::OnFocusMovedBetweenBrowsers(TabParent* aBlur,
                                                   TabParent* aFocus) {
   MOZ_ASSERT(aBlur != aFocus);
+  MOZ_ASSERT(XRE_IsParentProcess());
+
+  nsCOMPtr<nsIWidget> oldWidget = sWidget;
+  nsCOMPtr<nsIWidget> newWidget = aFocus ? aFocus->GetWidget() : nullptr;
+  // In the chrome-process case, we'll get sWidget from a PresShell later.
+  sWidget = newWidget;
+  if (oldWidget && sTextCompositions) {
+    RefPtr<TextComposition> composition =
+        sTextCompositions->GetCompositionFor(oldWidget);
+    if (composition) {
+      MOZ_LOG(
+          sISMLog, LogLevel::Debug,
+          ("  OnFocusMovedBetweenBrowsers(), requesting to commit "
+           "composition to "
+           "the (previous) focused widget (would request=%s)",
+           GetBoolName(
+               !oldWidget->IMENotificationRequestsRef().WantDuringDeactive())));
+      NotifyIME(REQUEST_TO_COMMIT_COMPOSITION, oldWidget,
+                composition->GetTabParent());
+    }
+  }
+
+  if (aBlur) {
+    MOZ_LOG(sISMLog, LogLevel::Debug,
+            ("  OnFocusMovedBetweenBrowsers(), notifying previous "
+             "focused child process of parent process or another child process "
+             "getting focus"));
+    Unused << aBlur->SendStopIMEStateManagement();
+  }
+
+  if (sActiveIMEContentObserver) {
+    DestroyIMEContentObserver();
+  }
+
+  if (sFocusedIMEWidget) {
+    // sFocusedIMETabParent can be null, if IME focus hasn't been
+    // taken before TabParent blur.
+    // aBlur can be null when keyboard focus moves not actually
+    // between tabs but an open menu is involved.
+    MOZ_ASSERT(!sFocusedIMETabParent || !aBlur ||
+               (sFocusedIMETabParent == aBlur));
+    MOZ_LOG(sISMLog, LogLevel::Debug,
+            ("  OnFocusMovedBetweenBrowsers(), notifying IME of blur"));
+    NotifyIME(NOTIFY_IME_OF_BLUR, sFocusedIMEWidget, sFocusedIMETabParent);
+
+    MOZ_ASSERT(!sFocusedIMETabParent);
+    MOZ_ASSERT(!sFocusedIMEWidget);
+
+  } else {
+    MOZ_ASSERT(!sFocusedIMETabParent);
+  }
+
+  // We deliberely don't null out sContent or sPresContext here. When
+  // focus is in remote content, as far as layout in the chrome process
+  // is concerned, the corresponding content is the top-level XUL
+  // browser. Changes among out-of-process iframes don't change that,
+  // so dropping the pointer to the XUL browser upon such a change
+  // would break IME handling.
 }
 
 // static
@@ -212,7 +238,10 @@
     sWidget = nullptr;
   }
   if (sFocusedIMEWidget == aWidget) {
-    NotifyIMEOfBlurForChildProcess();
+    if (sFocusedIMETabParent) {
+      OnFocusMovedBetweenBrowsers(sFocusedIMETabParent, nullptr);
+      MOZ_ASSERT(!sFocusedIMETabParent);
+    }
     sFocusedIMEWidget = nullptr;
   }
   if (sActiveInputContextWidget == aWidget) {
@@ -222,53 +251,22 @@
 
 // static
 void IMEStateManager::StopIMEStateManagement() {
+  MOZ_ASSERT(XRE_IsContentProcess());
   MOZ_LOG(sISMLog, LogLevel::Info, ("StopIMEStateManagement()"));
 
   // NOTE: Don't set input context from here since this has already lost
   //       the rights to change input context.
 
   if (sTextCompositions && sPresContext) {
-    NotifyIME(REQUEST_TO_COMMIT_COMPOSITION, sPresContext, sActiveTabParent);
+    NotifyIME(REQUEST_TO_COMMIT_COMPOSITION, sPresContext, nullptr);
   }
   sActiveInputContextWidget = nullptr;
   sPresContext = nullptr;
   sContent = nullptr;
-  sActiveTabParent = nullptr;
   DestroyIMEContentObserver();
 }
 
 // static
-void IMEStateManager::NotifyIMEOfBlurForChildProcess() {
-  MOZ_LOG(sISMLog, LogLevel::Debug,
-          ("NotifyIMEOfBlurForChildProcess(), sFocusedIMETabParent=0x%p, "
-           "sFocusedIMEWidget=0x%p",
-           sFocusedIMETabParent.get(), sFocusedIMEWidget));
-
-  if (!sFocusedIMETabParent) {
-    MOZ_ASSERT(!sFocusedIMEWidget);
-    return;
-  }
-
-  MOZ_ASSERT(sFocusedIMEWidget);
-
-  if (MOZ_LOG_TEST(sISMLog, LogLevel::Debug) && sTextCompositions) {
-    RefPtr<TextComposition> composition =
-        sTextCompositions->GetCompositionFor(sFocusedIMEWidget);
-    if (composition) {
-      MOZ_LOG(
-          sISMLog, LogLevel::Debug,
-          ("  NotifyIMEOfBlurForChildProcess(), sFocusedIMEWidget still has "
-           "composition"));
-    }
-  }
-
-  NotifyIME(NOTIFY_IME_OF_BLUR, sFocusedIMEWidget, sFocusedIMETabParent);
-
-  MOZ_ASSERT(!sFocusedIMETabParent);
-  MOZ_ASSERT(!sFocusedIMEWidget);
-}
-
-// static
 void IMEStateManager::MaybeStartOffsetUpdatedInChild(nsIWidget* aWidget,
                                                      uint32_t aStartOffset) {
   if (NS_WARN_IF(!sTextCompositions)) {
@@ -343,13 +341,12 @@
     InputContextAction action(InputContextAction::CAUSE_UNKNOWN,
                               InputContextAction::LOST_FOCUS);
     InputContext::Origin origin =
-        sActiveTabParent ? InputContext::ORIGIN_CONTENT : sOrigin;
+        TabParent::GetFocused() ? InputContext::ORIGIN_CONTENT : sOrigin;
     SetIMEState(newState, nullptr, nullptr, sWidget, action, origin);
   }
   sWidget = nullptr;
   sContent = nullptr;
   sPresContext = nullptr;
-  sActiveTabParent = nullptr;
   return NS_OK;
 }
 
@@ -399,14 +396,13 @@
     InputContextAction action(InputContextAction::CAUSE_UNKNOWN,
                               InputContextAction::LOST_FOCUS);
     InputContext::Origin origin =
-        sActiveTabParent ? InputContext::ORIGIN_CONTENT : sOrigin;
+        TabParent::GetFocused() ? InputContext::ORIGIN_CONTENT : sOrigin;
     SetIMEState(newState, aPresContext, nullptr, sWidget, action, origin);
   }
 
   sWidget = nullptr;
   sContent = nullptr;
   sPresContext = nullptr;
-  sActiveTabParent = nullptr;
 
   return NS_OK;
 }
@@ -433,21 +429,22 @@
 nsresult IMEStateManager::OnChangeFocusInternal(nsPresContext* aPresContext,
                                                 nsIContent* aContent,
                                                 InputContextAction aAction) {
-  RefPtr<TabParent> newTabParent = TabParent::GetFrom(aContent);
+  bool remoteHasFocus =
+      TabParent::GetFrom(aContent) || BrowserBridgeChild::GetFrom(aContent);
 
   MOZ_LOG(sISMLog, LogLevel::Info,
           ("OnChangeFocusInternal(aPresContext=0x%p (available: %s), "
-           "aContent=0x%p (TabParent=0x%p), aAction={ mCause=%s, "
+           "aContent=0x%p (remote: %s), aAction={ mCause=%s, "
            "mFocusChange=%s }), "
            "sPresContext=0x%p (available: %s), sContent=0x%p, "
-           "sWidget=0x%p (available: %s), sActiveTabParent=0x%p, "
+           "sWidget=0x%p (available: %s), TabParent::GetFocused()=0x%p, "
            "sActiveIMEContentObserver=0x%p, sInstalledMenuKeyboardListener=%s",
            aPresContext, GetBoolName(CanHandleWith(aPresContext)), aContent,
-           newTabParent.get(), GetActionCauseName(aAction.mCause),
+           GetBoolName(remoteHasFocus), GetActionCauseName(aAction.mCause),
            GetActionFocusChangeName(aAction.mFocusChange), sPresContext.get(),
            GetBoolName(CanHandleWith(sPresContext)), sContent.get(), sWidget,
            GetBoolName(sWidget && !sWidget->Destroyed()),
-           sActiveTabParent.get(), sActiveIMEContentObserver.get(),
+           TabParent::GetFocused(), sActiveIMEContentObserver.get(),
            GetBoolName(sInstalledMenuKeyboardListener)));
 
   // If new aPresShell has been destroyed, this should handle the focus change
@@ -465,7 +462,7 @@
       aPresContext ? aPresContext->GetRootWidget() : nullptr;
   bool focusActuallyChanging =
       (sContent != aContent || sPresContext != aPresContext ||
-       oldWidget != newWidget || sActiveTabParent != newTabParent);
+       oldWidget != newWidget || remoteHasFocus);
 
   // If old widget has composition, we may need to commit composition since
   // a native IME context is shared on all editors on some widgets or all
@@ -493,16 +490,10 @@
   }
 
   if (sActiveIMEContentObserver) {
-    // If there is active IMEContentObserver, it means that focused content was
-    // in this process.  So, if a tab parent gets focus, it means that the
-    // focused editor in this process is being blurred.
-    if (newTabParent) {
-      DestroyIMEContentObserver();
-    }
-    // If the process is being inactivated, then, IMEContentObserver should
-    // stop observing the contents unless native IME requests to keep
-    // composition even during deactivated.
-    else if (!aPresContext) {
+    MOZ_ASSERT(!remoteHasFocus,
+               "IMEContentObserver should have been destroyed by "
+               "OnFocusMovedBetweenBrowsers.");
+    if (!aPresContext) {
       if (!sActiveIMEContentObserver->KeepAliveDuringDeactive()) {
         DestroyIMEContentObserver();
       }
@@ -513,29 +504,6 @@
     else if (!sActiveIMEContentObserver->IsManaging(aPresContext, aContent)) {
       DestroyIMEContentObserver();
     }
-  } else {
-    // If there is no active IMEContentObserver, it means that focused content
-    // may be in another process.
-
-    // If focus is moving from current focused remote process to different
-    // process while the process has IME focus too, we need to notify IME of
-    // blur here because it may be too late the blur notification to reach
-    // this process especially when closing active window.
-    // However, don't send blur if we're being deactivated and IME wants to
-    // keep composition during deactive because notifying blur will commit
-    // or cancel composition.
-    if (sFocusedIMETabParent && sFocusedIMEWidget &&
-        (aPresContext || !sFocusedIMEWidget->IMENotificationRequestsRef()
-                              .WantDuringDeactive()) &&
-        !IsSameProcess(sFocusedIMETabParent, newTabParent)) {
-      MOZ_LOG(
-          sISMLog, LogLevel::Info,
-          ("  OnChangeFocusInternal(), notifying IME of blur of previous "
-           "focused "
-           "remote process because it may be too late actual notification to "
-           "reach this process"));
-      NotifyIMEOfBlurForChildProcess();
-    }
   }
 
   if (!aPresContext) {
@@ -545,14 +513,6 @@
     return NS_OK;
   }
 
-  if (sActiveTabParent && !IsSameProcess(sActiveTabParent, newTabParent)) {
-    MOZ_LOG(sISMLog, LogLevel::Debug,
-            ("  OnChangeFocusInternal(), notifying previous "
-             "focused child process of parent process or another child process "
-             "getting focus"));
-    Unused << sActiveTabParent->SendStopIMEStateManagement();
-  }
-
   if (NS_WARN_IF(!newWidget)) {
     MOZ_LOG(sISMLog, LogLevel::Error,
             ("  OnChangeFocusInternal(), FAILED due to "
@@ -567,12 +527,11 @@
   // If a child process has focus, we should disable IME state until the child
   // process actually gets focus because if user types keys before that they
   // are handled by IME.
-  IMEState newState = newTabParent ? IMEState(IMEState::DISABLED)
-                                   : GetNewIMEState(aPresContext, aContent);
+  IMEState newState = remoteHasFocus ? IMEState(IMEState::DISABLED)
+                                     : GetNewIMEState(aPresContext, aContent);
   bool setIMEState = true;
 
-  if (newTabParent) {
-    MOZ_ASSERT(XRE_IsParentProcess());
+  if (remoteHasFocus && XRE_IsParentProcess()) {
     if (aAction.mFocusChange == InputContextAction::MENU_GOT_PSEUDO_FOCUS) {
       // If menu keyboard listener is installed, we need to disable IME now.
       setIMEState = true;
@@ -654,7 +613,7 @@
                                       : InputContextAction::LOST_FOCUS;
     }
 
-    if (newTabParent && HasActiveChildSetInputContext() &&
+    if (remoteHasFocus && HasActiveChildSetInputContext() &&
         aAction.mFocusChange == InputContextAction::MENU_LOST_PSEUDO_FOCUS) {
       // Restore the input context in the active remote process when
       // menu keyboard listener is uninstalled and active remote tab has
@@ -663,11 +622,10 @@
     } else {
       // Update IME state for new focus widget
       SetIMEState(newState, aPresContext, aContent, newWidget, aAction,
-                  newTabParent ? InputContext::ORIGIN_CONTENT : sOrigin);
+                  remoteHasFocus ? InputContext::ORIGIN_CONTENT : sOrigin);
     }
   }
 
-  sActiveTabParent = newTabParent;
   sPresContext = aPresContext;
   sContent = aContent;
 
@@ -694,12 +652,12 @@
   MOZ_LOG(
       sISMLog, LogLevel::Info,
       ("OnInstalledMenuKeyboardListener(aInstalling=%s), "
-       "sInstalledMenuKeyboardListener=%s, sActiveTabParent=0x%p, "
+       "sInstalledMenuKeyboardListener=%s, TabParent::GetFocused()=0x%p, "
        "sActiveChildInputContext={ mIMEState={ mEnabled=%s, mOpen=%s }, "
        "mHTMLInputType=\"%s\", mHTMLInputInputmode=\"%s\", mActionHint=\"%s\", "
        "mInPrivateBrowsing=%s }",
        GetBoolName(aInstalling), GetBoolName(sInstalledMenuKeyboardListener),
-       sActiveTabParent.get(),
+       TabParent::GetFocused(),
        GetIMEStateEnabledName(sActiveChildInputContext.mIMEState.mEnabled),
        GetIMEStateSetOpenName(sActiveChildInputContext.mIMEState.mOpen),
        NS_ConvertUTF16toUTF8(sActiveChildInputContext.mHTMLInputType).get(),
@@ -714,6 +672,14 @@
                             aInstalling
                                 ? InputContextAction::MENU_GOT_PSEUDO_FOCUS
                                 : InputContextAction::MENU_LOST_PSEUDO_FOCUS);
+  TabParent* focused = TabParent::GetFocused();
+  if (focused) {
+    if (aInstalling) {
+      OnFocusMovedBetweenBrowsers(focused, nullptr);
+    } else {
+      OnFocusMovedBetweenBrowsers(nullptr, focused);
+    }
+  }
   OnChangeFocusInternal(sPresContext, sContent, action);
 }
 
@@ -1156,7 +1122,7 @@
        "mHTMLInputType=\"%s\", mHTMLInputInputmode=\"%s\", mActionHint=\"%s\", "
        "mInPrivateBrowsing=%s }, aAction={ mCause=%s, mAction=%s }), "
        "sPresContext=0x%p (available: %s), sWidget=0x%p (available: %s), "
-       "sActiveTabParent=0x%p, sInstalledMenuKeyboardListener=%s",
+       "TabParent::GetFocused()=0x%p, sInstalledMenuKeyboardListener=%s",
        aTabParent, GetIMEStateEnabledName(aInputContext.mIMEState.mEnabled),
        GetIMEStateSetOpenName(aInputContext.mIMEState.mOpen),
        NS_ConvertUTF16toUTF8(aInputContext.mHTMLInputType).get(),
@@ -1166,10 +1132,10 @@
        GetActionCauseName(aAction.mCause),
        GetActionFocusChangeName(aAction.mFocusChange), sPresContext.get(),
        GetBoolName(CanHandleWith(sPresContext)), sWidget,
-       GetBoolName(sWidget && !sWidget->Destroyed()), sActiveTabParent.get(),
+       GetBoolName(sWidget && !sWidget->Destroyed()), TabParent::GetFocused(),
        GetBoolName(sInstalledMenuKeyboardListener)));
 
-  if (aTabParent != sActiveTabParent) {
+  if (aTabParent != TabParent::GetFocused()) {
     MOZ_LOG(sISMLog, LogLevel::Error,
             ("  SetInputContextForChildProcess(), FAILED, "
              "because non-focused tab parent tries to set input context"));
@@ -1344,21 +1310,21 @@
 void IMEStateManager::SetInputContext(nsIWidget* aWidget,
                                       const InputContext& aInputContext,
                                       const InputContextAction& aAction) {
-  MOZ_LOG(
-      sISMLog, LogLevel::Info,
-      ("SetInputContext(aWidget=0x%p, aInputContext={ "
-       "mIMEState={ mEnabled=%s, mOpen=%s }, mHTMLInputType=\"%s\", "
-       "mHTMLInputInputmode=\"%s\", mActionHint=\"%s\", "
-       "mInPrivateBrowsing=%s }, "
-       "aAction={ mCause=%s, mAction=%s }), sActiveTabParent=0x%p",
-       aWidget, GetIMEStateEnabledName(aInputContext.mIMEState.mEnabled),
-       GetIMEStateSetOpenName(aInputContext.mIMEState.mOpen),
-       NS_ConvertUTF16toUTF8(aInputContext.mHTMLInputType).get(),
-       NS_ConvertUTF16toUTF8(aInputContext.mHTMLInputInputmode).get(),
-       NS_ConvertUTF16toUTF8(aInputContext.mActionHint).get(),
-       GetBoolName(aInputContext.mInPrivateBrowsing),
-       GetActionCauseName(aAction.mCause),
-       GetActionFocusChangeName(aAction.mFocusChange), sActiveTabParent.get()));
+  MOZ_LOG(sISMLog, LogLevel::Info,
+          ("SetInputContext(aWidget=0x%p, aInputContext={ "
+           "mIMEState={ mEnabled=%s, mOpen=%s }, mHTMLInputType=\"%s\", "
+           "mHTMLInputInputmode=\"%s\", mActionHint=\"%s\", "
+           "mInPrivateBrowsing=%s }, "
+           "aAction={ mCause=%s, mAction=%s }), TabParent::GetFocused()=0x%p",
+           aWidget, GetIMEStateEnabledName(aInputContext.mIMEState.mEnabled),
+           GetIMEStateSetOpenName(aInputContext.mIMEState.mOpen),
+           NS_ConvertUTF16toUTF8(aInputContext.mHTMLInputType).get(),
+           NS_ConvertUTF16toUTF8(aInputContext.mHTMLInputInputmode).get(),
+           NS_ConvertUTF16toUTF8(aInputContext.mActionHint).get(),
+           GetBoolName(aInputContext.mInPrivateBrowsing),
+           GetActionCauseName(aAction.mCause),
+           GetActionFocusChangeName(aAction.mFocusChange),
+           TabParent::GetFocused()));
 
   MOZ_RELEASE_ASSERT(aWidget);
 
@@ -1579,13 +1545,13 @@
       sISMLog, LogLevel::Info,
       ("NotifyIME(aNotification={ mMessage=%s }, "
        "aWidget=0x%p, aTabParent=0x%p), sFocusedIMEWidget=0x%p, "
-       "sActiveTabParent=0x%p, sFocusedIMETabParent=0x%p, "
-       "IsSameProcess(aTabParent, sActiveTabParent)=%s, "
-       "IsSameProcess(aTabParent, sFocusedIMETabParent)=%s",
+       "TabParent::GetFocused()=0x%p, sFocusedIMETabParent=0x%p, "
+       "aTabParent == TabParent::GetFocused()=%s, "
+       "aTabParent == sFocusedIMETabParent=%s",
        ToChar(aNotification.mMessage), aWidget, aTabParent, sFocusedIMEWidget,
-       sActiveTabParent.get(), sFocusedIMETabParent.get(),
-       GetBoolName(IsSameProcess(aTabParent, sActiveTabParent)),
-       GetBoolName(IsSameProcess(aTabParent, sFocusedIMETabParent))));
+       TabParent::GetFocused(), sFocusedIMETabParent.get(),
+       GetBoolName(aTabParent == TabParent::GetFocused()),
+       GetBoolName(aTabParent == sFocusedIMETabParent)));
 
   if (NS_WARN_IF(!aWidget)) {
     MOZ_LOG(sISMLog, LogLevel::Error,
@@ -1595,46 +1561,53 @@
 
   switch (aNotification.mMessage) {
     case NOTIFY_IME_OF_FOCUS: {
-      // If focus notification comes from a remote process which already lost
+      // If focus notification comes from a remote browser which already lost
       // focus, we shouldn't accept the focus notification.  Then, following
       // notifications from the process will be ignored.
-      if (NS_WARN_IF(!IsSameProcess(aTabParent, sActiveTabParent))) {
-        MOZ_ASSERT(
-            aTabParent,
-            "Why was the input context initialized for a remote process but "
-            "does this process get IME focus?");
-        MOZ_LOG(
-            sISMLog, LogLevel::Warning,
-            ("  NotifyIME(), WARNING, the received focus notification is "
-             "ignored "
-             "because input context was initialized for %s, perhaps, it came "
-             "from a busy remote process",
-             sActiveTabParent ? "another remote process" : "current process"));
+      if (aTabParent != TabParent::GetFocused()) {
+        MOZ_LOG(sISMLog, LogLevel::Warning,
+                ("  NotifyIME(), WARNING, the received focus notification is "
+                 "ignored, because its associated TabParent did not match"
+                 "the focused TabParent."));
         return NS_OK;
       }
-      // If there is pending blur notification for current focused IME,
-      // we should notify IME of blur by ourselves.  Then, we should ignore
-      // following notifications coming from the process.
+      // If IME focus is already set, first blur the currently-focused
+      // IME widget
       if (sFocusedIMEWidget) {
+        // XXX Why don't we first request the previously-focused IME
+        // widget to commit the composition?
         MOZ_ASSERT(
             sFocusedIMETabParent || aTabParent,
             "This case shouldn't be caused by focus move in this process");
+        MOZ_LOG(sISMLog, LogLevel::Warning,
+                ("  NotifyIME(), WARNING, received focus notification with ")
+                 "non-null sFocusedIMEWidget. How come "
+                 "OnFocusMovedBetweenBrowsers did not blur it already?");
         nsCOMPtr<nsIWidget> focusedIMEWidget(sFocusedIMEWidget);
         sFocusedIMEWidget = nullptr;
         sFocusedIMETabParent = nullptr;
         focusedIMEWidget->NotifyIME(IMENotification(NOTIFY_IME_OF_BLUR));
       }
+#ifdef DEBUG
+      if (aTabParent) {
+        nsCOMPtr<nsIWidget> tabParentWidget = aTabParent->GetWidget();
+        MOZ_ASSERT(tabParentWidget == aWidget);
+      }
+#endif
       sFocusedIMETabParent = aTabParent;
       sFocusedIMEWidget = aWidget;
       nsCOMPtr<nsIWidget> widget(aWidget);
+      MOZ_LOG(
+          sISMLog, LogLevel::Info,
+          ("  NotifyIME(), about to call widget->NotifyIME() for IME focus"));
       return widget->NotifyIME(aNotification);
     }
     case NOTIFY_IME_OF_BLUR: {
-      if (!IsSameProcess(aTabParent, sFocusedIMETabParent)) {
+      if (aTabParent != sFocusedIMETabParent) {
         MOZ_LOG(sISMLog, LogLevel::Warning,
                 ("  NotifyIME(), WARNING, the received blur notification is "
                  "ignored "
-                 "because it's not from current focused IME process"));
+                 "because it's not from current focused IME browser"));
         return NS_OK;
       }
       if (!sFocusedIMEWidget) {
@@ -1661,11 +1634,11 @@
     case NOTIFY_IME_OF_POSITION_CHANGE:
     case NOTIFY_IME_OF_MOUSE_BUTTON_EVENT:
     case NOTIFY_IME_OF_COMPOSITION_EVENT_HANDLED: {
-      if (!IsSameProcess(aTabParent, sFocusedIMETabParent)) {
+      if (aTabParent != sFocusedIMETabParent) {
         MOZ_LOG(
             sISMLog, LogLevel::Warning,
             ("  NotifyIME(), WARNING, the received content change notification "
-             "is ignored because it's not from current focused IME process"));
+             "is ignored because it's not from current focused IME browser"));
         return NS_OK;
       }
       if (!sFocusedIMEWidget) {
@@ -1707,11 +1680,11 @@
     return NS_OK;
   }
 
-  if (!IsSameProcess(aTabParent, composition->GetTabParent())) {
+  if (aTabParent != composition->GetTabParent()) {
     MOZ_LOG(
         sISMLog, LogLevel::Warning,
         ("  NotifyIME(), WARNING, the request to IME is ignored because "
-         "it does not come from the remote process which has the composition "
+         "it does not come from the remote browser which has the composition "
          "on aWidget"));
     return NS_OK;
   }